core/src/main/scala/org/apache/spark/Accumulable.scala

```diff
@@ -146,6 +212,32 @@ class Accumulable[R, T] private (
   def merge(term: R) { value_ = param.addInPlace(value_, term)}
 
   /**
+   * Merge in pending updates for ac consistent accumulators or merge accumulated values for
+   * regular accumulators. This is only called on the driver when merging task results together.
+   */
+  private[spark] def internalMerge(term: Any) {
+    if (!consistent) {
+      merge(term.asInstanceOf[R])
+    } else {
+      mergePending(term.asInstanceOf[mutable.HashMap[(Int, Int, Int), R]])
+    }
+  }
+
+  /**
+   * Merge another Accumulable's pending updates, checks to make sure that each pending update has
+   * not already been processed before updating.
+   */
+  private[spark] def mergePending(term: mutable.HashMap[(Int, Int, Int), R]) = {
+    term.foreach{case ((rddId, shuffleId, splitId), v) =>
+      val splits = processed.getOrElseUpdate((rddId, shuffleId), new mutable.BitSet())
+      if (!splits.contains(splitId)) {
+        splits += splitId
+        value_ = param.addInPlace(value_, v)
+      }
```

Comment:
I don't think you need both `completed` and `processed` -- they seem to be doing more or less the same thing.  You could change this to:

```scala
    term.foreach{case (partitionKey, v) =>
      if (!completed.contains(partitionKey)) {
        completed += partitionKey
        value_ = param.addInPlace(value_, v)
      }
    }
```

If I understand correctly, there is a bit of logical distinction between the two -- `processed` was being used on the driver, that is across multiple tasks, to track what had been accumulated and what hadn't.  `completed`, OTOH, was being used on the executors, to track the updates coming from *one task*.  Typically it would only contain a single entry, for the one `(rdd, shuffle, partition)` tuple that was being used, since tasks *usually* only compute one partition, but that isn't true when there is a coalesce involved.

If that explanation sounds correct, its probably best to put it into a comment somewhere, and I'd still say that you can eliminate `processed` and just explain how `completed` will get used in two different ways on the executors and on the driver.